src/main/java/org/apache/sysml/runtime/matrix/data/LibMatrixDNN.java

```diff
@@ -0,0 +1,564 @@
+				start_index_h = Math.max(start_index_h, 0);
+				start_index_w = Math.max(start_index_w, 0);
+				int maxIndex = n*params.C*params.H*params.W + c*params.H*params.W +  start_index_h*params.W + start_index_w; 
+				double maxVal = -Double.MAX_VALUE; 
+
+
+				double currDoutVal = -1;
+				for (int h = start_index_h; h < end_index_h; h++) {
+					for (int w = start_index_w; w < end_index_w; w++) {
+						if(inputArray != null)
+							currDoutVal = inputArray[n*params.C*params.H*params.W + c*params.H*params.W +  h*params.W + w];
+						else
+							currDoutVal = params.input1.quickGetValue(n, c*params.H*params.W + h*params.W + w);
+
+						if(maxVal < currDoutVal) {
+							maxIndex = n*params.C*params.H*params.W + c*params.H*params.W +  h*params.W + w;
+							maxVal = currDoutVal;
+						}
+					}
+				}
+
+				double inVal = -1;
+				if(doutArray != null)
+					inVal = doutArray[n*params.C*params.P*params.Q + c*params.P*params.Q +  p * params.Q + q];
+				else
+					inVal = params.input2.quickGetValue(n, c*params.P*params.Q +  p * params.Q + q);
+
+				// synchronized(this) {
+					outputArray[maxIndex] += inVal;
+				// }
+			}
+		}
+	}
+
+	public static void maxpooling(MatrixBlock input, MatrixBlock outputBlock, ConvolutionParameters params) throws DMLRuntimeException {
+		params.input1 = input;
+		params.output = outputBlock;
+		
+		if(input.getNumColumns() != params.C*params.H*params.W || input.getNumRows() != params.N) {
+			throw new DMLRuntimeException("Incorrect input dimensions in maxpooling:" + input.getNumRows() + " " + input.getNumColumns() + " " + params.N + " " + params.K*params.P*params.Q);
+		}
+		
+		int constrainedNumThreads = OptimizerUtils.getConstrainedNumThreads(params.numThreads);
+		if(!ALLOW_MULTI_THREADED_OPS || constrainedNumThreads <= 1) {
+			for (int n = 0; n < params.N; n++) {
+				for (int c = 0; c < params.C; c++) {
+					doPooling(n, c, params);
+				}
+			}
+		}
+		else {
+			runParallelConvTask(constrainedNumThreads, params.C, TaskType.MaxPooling_Forward, params);
+		}	
+	}
+
+	private static void doPooling(int n, int c, ConvolutionParameters params) {
+		double [] inputArray = null;
+		if (!params.input1.isInSparseFormat())
+			inputArray = params.input1.getDenseBlock();
+		double [] outputArray = null;
+		if (!params.output.isInSparseFormat())
+			outputArray = params.output.getDenseBlock();
+		
+		for (int p = 0; p < params.P; p++) {
+			for (int q = 0; q < params.Q; q++) {
+				int start_index_h = p * params.stride_h - params.pad_h;
+				int start_index_w = q * params.stride_w - params.pad_w;
+				int end_index_h = Math.min(start_index_h + params.R, params.H);
+				int end_index_w = Math.min(start_index_w + params.S, params.W);
+				start_index_h = Math.max(start_index_h, 0);
+				start_index_w = Math.max(start_index_w, 0);
+				int out_index = n*params.C*params.P*params.Q + c*params.P*params.Q +  p * params.Q + q;
+				outputArray[out_index] = 0;
```

Comment:
If we update this to `outputArray[out_index] = -Double.MAX_VALUE;`, patches with all negative values will produce the correct result.